internal/resources/metal/vlans/models.go

```diff
@@ -0,0 +1,81 @@
+package vlans
+
+import (
+	"context"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/packethost/packngo"
+	"strings"
+)
+
+type DataSourceModel struct {
+	ID                 types.String `tfsdk:"id"`
+	ProjectID          types.String `tfsdk:"project_id"`
+	VlanID             types.String `tfsdk:"vlan_id"`
+	Vxlan              types.Int64  `tfsdk:"vxlan"`
+	Facility           types.String `tfsdk:"facility"`
+	Metro              types.String `tfsdk:"metro"`
+	Description        types.String `tfsdk:"description"`
+	AssignedDevicesIds types.List   `tfsdk:"assigned_devices_ids"`
+}
+
+func (m *DataSourceModel) parse(vlan *packngo.VirtualNetwork) diag.Diagnostics {
+	m.ID = types.StringValue(vlan.ID)
+
+	if vlan.Project.ID != "" {
+		m.ProjectID = types.StringValue(vlan.Project.ID)
+	}
+
+	m.VlanID = types.StringValue(vlan.ID)
+
+	m.Facility = types.StringNull()
+	if vlan.FacilityCode != "" {
+		m.Facility = types.StringValue(vlan.FacilityCode)
+	}
+
+	m.Description = types.StringValue(vlan.Description)
+	m.Vxlan = types.Int64Value(int64(vlan.VXLAN))
+
+	// version of this resource. StateFunc doesn't exist in terraform and it requires implementation
+	// of bespoke logic before storing state. To ensure backward compatibility we ignore lower/upper
+	// case diff for now, but we may want to require input upper case
+	if !strings.EqualFold(m.Metro.ValueString(), vlan.MetroCode) {
```

Comment:
The API seems to return `null` for `metro` and `metro_code` when a VLAN was created in a facility, even if you `include=metro`; for a VLAN created in a facility we have to read the metro code from `vlan.Facility.Metro.Code`

Reply:
Also I think we need a test for the metro casing behavior; I suspect that an `EqualFold` check while parsing API responses isn't enough to suppress the diff, but I don't know.  The test would:
- create a vlan with `metro = "da"`
- change to `metro = "DA"`
- expect an empty plan